sdk/identity/azure-identity/azure/identity/aio/_internal/msal_credentials.py

```diff
@@ -0,0 +1,134 @@
+because MSAL has no async support.
+
+This is what happens when a new token is required to service a ``get_token`` call:
+  1. The credential submits the appropriate MSAL token acquisition method to an Executor, receiving a future.
+  2. The credential awaits the future's completion, freeing the caller's event loop until the MSAL method returns.
+  3. MSAL calls into the transport adapter in whatever context the Executor provided (a non-main thread by default).
+  4. The adapter schedules an async pipeline run on the ``get_token`` caller's loop, and blocks on the result.
+  5. The adapter wraps the pipeline response in the shape MSAL expects (requests.Response) and returns it.
+  6. The MSAL token acquisition method returns; the future obtained in step 1 completes.
+  7. The credential packages the future's result as an AccessToken and returns it to the caller.
+
+The upshot is, this looks like an ordinary ``get_token`` coroutine to whomever holds the credential. The async
+transport looks like an ordinary Requests session to MSAL.
+"""
+
+import abc
+import asyncio
+import functools
+import time
+from typing import TYPE_CHECKING
+from unittest import mock
+
+if TYPE_CHECKING:
+    # pylint:disable=unused-import
+    from typing import Dict, Optional, Type, Union
+    from concurrent.futures import Executor
+
+from azure.core.credentials import AccessToken
+from azure.core.exceptions import ClientAuthenticationError
+import msal
+
+from .msal_transport_adapter import AsyncMsalTransportAdapter
+
+
+class MsalCredential(abc.ABC):
+    """Base class for async credentials wrapping MSAL applications"""
+
+    def __init__(
+        self, client_id: str, authority: str, client_credential: "Optional[Union[str, Dict[str,str]]]", **kwargs: "Any"
+    ) -> None:
+        self._authority = authority
+        self._client_credential = client_credential
+        self._client_id = client_id
+
+        self._lock = asyncio.Lock()
+        self._adapter = kwargs.pop("msal_adapter", None) or AsyncMsalTransportAdapter(**kwargs)
+
+        # postpone creating the wrapped application because its initializer uses the network
+        self._msal_app = None  # type: Optional[msal.ClientApplication]
+
+    @abc.abstractmethod
+    def get_token(self, *scopes: str) -> AccessToken:
+        pass
+
+    @abc.abstractmethod
+    def _get_app(self) -> msal.ClientApplication:
+        pass
+
+    async def _create_app(
+        self, cls: "Type[msal.ClientApplication]", executor: "Optional[Executor]" = None
+    ) -> msal.ClientApplication:
+        """Creates an MSAL application, patching msal.authority to use an azure-core pipeline during tenant discovery"""
+
+        if self._msal_app:
+            return self._msal_app
+
+        initializer = functools.partial(
+            cls,
+            client_id=self._client_id,
+            client_credential=self._client_credential,
+            authority=self._authority,
+            validate_authority=False,
```

Comment:
Why are we disabling authority validation?

Reply:
"Authority validation" here is GETing the tenant discovery endpoint of an AAD instance unknown to MSAL. I opted out of it to prevent that request. This is tantamount to assuming the instance behaves like e.g. login.microsoftonline.com, and in fact is presently irrelevant because we're using a hard-coded authority known to MSAL.